evm_loader/program/src/instruction/transaction_step.rs

```diff
@@ -14,56 +13,75 @@ pub fn do_begin<'a>(
+
+    // Increment origin nonce in the first iteration
+    // This allows us to run multiple iterative transactions from the same sender in parallel
+    // These transactions are guaranteed to start in a correct sequence
+    // BUT they finalize in an undefined order
+    let origin_nonce = origin_account.nonce();
+    let trx_nonce = storage.trx().nonce();
+    if origin_nonce != trx_nonce {
+        let error = Error::InvalidTransactionNonce(origin, origin_nonce, trx_nonce);
+        return Err(error);
+    }
+    origin_account.increment_nonce()?;
 
     // Burn `gas_limit` tokens from the origin account
     // Later we will mint them to the operator
-    let mut origin_balance = accounts.create_balance_account(origin, storage.trx_chain_id())?;
-    origin_balance.burn(storage.gas_limit_in_tokens()?)?;
+    // Remaining tokens are returned to the origin in the last iteration
+    let gas_limit_in_tokens = storage.trx().gas_limit_in_tokens()?;
+    origin_account.burn(gas_limit_in_tokens)?;
+
+    // Initialize EVM and serialize it to the Holder
+    let mut backend = ExecutorState::new(&account_storage);
+    let evm = Machine::new(storage.trx(), origin, &mut backend)?;
 
     serialize_evm_state(&mut storage, &backend, &evm)?;
-    finalize(0, storage, accounts, None, gasometer)
+    finalize(0, storage, account_storage, None, gasometer)
 }
 
 pub fn do_continue<'a>(
     step_count: u64,
     accounts: AccountsDB<'a>,
     mut storage: StateAccount<'a>,
     gasometer: Gasometer,
+    reset: bool,
 ) -> Result<()> {
     debug_print!("do_continue");
 
-    if (step_count < EVM_STEPS_MIN) && (storage.trx_gas_price() > 0) {
+    if (step_count < EVM_STEPS_MIN) && (storage.trx().gas_price() > 0) {
         return Err(Error::Custom(format!(
             "Step limit {step_count} below minimum {EVM_STEPS_MIN}"
         )));
     }
 
     let account_storage = ProgramAccountStorage::new(accounts)?;
-    let (mut backend, mut evm) = deserialize_evm_state(&storage, &account_storage)?;
+    let (mut backend, mut evm) = if reset {
+        let mut backend = ExecutorState::new(&account_storage);
+        let evm = Machine::new(storage.trx(), storage.trx_origin(), &mut backend)?;
+        (backend, evm)
+    } else {
+        deserialize_evm_state(&storage, &account_storage)?
+    };
 
-    let (result, steps_executed) = {
-        match backend.exit_status() {
-            Some(status) => (status.clone(), 0_u64),
-            None => evm.execute(step_count, &mut backend)?,
-        }
+    let (result, steps_executed) = match backend.exit_status() {
+        Some(status) => (status.clone(), 0_u64),
+        None => evm.execute(step_count, &mut backend)?,
     };
 
     if (result != ExitStatus::StepLimit) && (steps_executed > 0) {
         backend.set_exit_status(result.clone());
     }
 
-    if steps_executed > 0 {
-        serialize_evm_state(&mut storage, &backend, &evm)?;
-    }
+    serialize_evm_state(&mut storage, &backend, &evm)?;
```

Comment:
We can reset the state to the initial state and execute zero steps at the same time.
Since state changed, it would need to be serialized to the Holder